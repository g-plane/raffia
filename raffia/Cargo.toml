[package]
name = "raffia"
version = "0.1.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "One parser for CSS/SCSS/Sass/Less."
repository = "https://github.com/g-plane/raffia"
license = "MIT"

[dependencies]
raffia_macro = { path = "../raffia_macro", version = "0.1" }
serde = { version = "1.0", features = ["derive"], optional = true }
smallvec = { version = "1.9", features = ["const_generics", "serde"] }

[dev-dependencies]
codespan-reporting = "0.11"
criterion = "0.3"
insta = { version = "1.19", features = ["glob", "ron"] }
lightningcss = "1.0.0-alpha.35"
serde_json = "1.0"
swc_common = "0.29.0"
swc_css_ast = "0.114.0"
swc_css_parser = "0.123.0"

[features]
serialize = ["serde"]

[[bench]]
name = "parse_selector"
harness = false

[[bench]]
name = "parse_stylesheet"
harness = false
