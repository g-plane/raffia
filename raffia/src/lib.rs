//! Raffia is a parser can parse CSS, SCSS, Sass (indented syntax) and Less.
//!
//! ## Basic Usage
//!
//! This crate provides a simple API to get started.
//!
//! First, create a parser, give it the source code and specify the syntax,
//! then call the [`parse`](Parser::parse) method:
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css); // syntax can also be `Scss`, `Sass` or `Less`
//! let result = parser.parse::<Stylesheet>();
//! match result {
//!     Ok(ast) => {
//!         // parsed successfully
//!         println!("{:#?}", ast);
//!     }
//!     Err(error) => {
//!         // it failed, error message and position can be accessed via `error`
//!         println!("{:#?}", error);
//!     }
//! }
//! ```
//!
//! ## Advanced Usage
//!
//! ### More Parser Options
//!
//! If you need to control more parser options, you can use [`ParserBuilder`].
//!
//! For example, to collect comments:
//!
//! ```rust
//! use raffia::ParserBuilder;
//!
//! let mut comments = vec![];
//! let builder = ParserBuilder::new("/* comment */ a {}").comments(&mut comments);
//! let mut parser = builder.build();
//! ```
//!
//! By default, syntax is CSS when using parser builder. You can customize it:
//!
//! ```rust
//! use raffia::{ParserBuilder, Syntax};
//!
//! let builder = ParserBuilder::new("a {}").syntax(Syntax::Scss);
//! ```
//!
//! ### Parse Partial Structure
//!
//! Sometimes you don't want to parse a full stylesheet.
//! Say you only need to parse a qualified rule or even a single declaration.
//! All you need to do is to update the generics of the [`parse`](Parser::parse) method.
//!
//! ```rust
//! use raffia::{ast::QualifiedRule, Parser, Syntax};
//!
//! let mut parser = Parser::new("a {}", Syntax::Css);
//! parser.parse::<QualifiedRule>();
//! ```
//!
//! and
//!
//! ```rust
//! use raffia::{ast::Declaration, Parser, Syntax};
//!
//! let mut parser = Parser::new("color: green", Syntax::Css);
//! parser.parse::<Declaration>();
//! ```
//!
//! Not all AST nodes support the usage above;
//! technically, those nodes that implement [`Parse`] trait are supported.
//!
//! ### Retrieve Recoverable Errors
//!
//! There may be some recoverable errors which doesn't affect on producing AST.
//! To retrieve those errors, use [`recoverable_errors`](Parser::recoverable_errors).
//!
//! ```rust
//! use raffia::{ast::Stylesheet, Parser, Syntax};
//!
//! let mut parser = Parser::new("@keyframes kf { invalid {} }", Syntax::Css);
//! let result = parser.parse::<Stylesheet>();
//! assert!(result.is_ok());
//! println!("{:?}", parser.recoverable_errors());
//! ```
//!
//! ## Serialization
//!
//! Produced AST can be serialized by Serde, but this feature is disabled by default.
//! You need to enable feature `serialize` manually:
//!
//! ```toml
//! raffia = { version = "*", features = ["serialize"] }
//! ```
//!
//! Then you can pass AST to Serde.
//!
//! Note that Raffia only supports serialization. Deserialization isn't supported.

pub use config::{ParserOptions, Syntax};
pub use parser::{Parse, Parser, ParserBuilder};
pub use pos::{Span, Spanned};
pub use span_ignored_eq::SpanIgnoredEq;
pub use tokenizer::token;

pub mod ast;
mod config;
pub mod error;
mod parser;
pub mod pos;
mod span_ignored_eq;
mod tokenizer;
mod util;
