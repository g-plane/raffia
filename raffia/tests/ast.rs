use codespan_reporting::{
    diagnostic::{Diagnostic, Label},
    files::SimpleFile,
    term::{self, termcolor::Buffer},
};
use insta::{assert_ron_snapshot, glob, Settings};
use raffia::{ast::Stylesheet, Parser, Syntax};
use std::fs;

#[test]
fn ast_snapshot() {
    glob!("ast/**/*.{css,scss,sass,less}", |path| {
        let code = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };
        let mut parser = Parser::new(&code, syntax);
        let ast = match parser.parse::<Stylesheet>() {
            Ok(ast) => {
                let recoverable_errors = parser.recoverable_errors();
                assert!(
                    recoverable_errors.is_empty(),
                    "'{}' has recoverable errors: {recoverable_errors:?}",
                    path.file_name().unwrap().to_str().unwrap(),
                );
                ast
            }
            Err(error) => {
                let file = SimpleFile::new(path.file_name().unwrap().to_str().unwrap(), &code);
                let diagnostic = Diagnostic::error()
                    .with_message(error.kind.to_string())
                    .with_labels(vec![Label::primary((), error.span.start..error.span.end)]);
                let mut buffer = Buffer::ansi();
                let config = term::Config::default();
                term::emit(&mut buffer, &config, &file, &diagnostic).unwrap();
                panic!("\n{}", String::from_utf8(buffer.into_inner()).unwrap());
            }
        };

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_ron_snapshot!(name, ast);
        });
    });
}
