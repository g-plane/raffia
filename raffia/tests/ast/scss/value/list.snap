---
source: raffia/tests/ast.rs
---
Stylesheet(
  type: "Stylesheet",
  statements: [
    SassVariableDeclaration(
      type: "SassVariableDeclaration",
      namespace: None,
      name: SassVariable(
        type: "SassVariable",
        name: Ident(
          type: "Ident",
          name: "mylist",
          raw: "mylist",
          span: Span(
            start: 1,
            end: 7,
          ),
        ),
        span: Span(
          start: 0,
          end: 7,
        ),
      ),
      value: SassList(
        type: "SassList",
        items: [
          Ident(
            type: "Ident",
            name: "alpha",
            raw: "alpha",
            span: Span(
              start: 10,
              end: 15,
            ),
          ),
          Ident(
            type: "Ident",
            name: "beta",
            raw: "beta",
            span: Span(
              start: 17,
              end: 21,
            ),
          ),
          Ident(
            type: "Ident",
            name: "gamma",
            raw: "gamma",
            span: Span(
              start: 23,
              end: 28,
            ),
          ),
        ],
        hasParens: true,
        separator: Comma,
        span: Span(
          start: 9,
          end: 31,
        ),
      ),
      important: None,
      overridable: false,
      forceGlobal: false,
      span: Span(
        start: 0,
        end: 32,
      ),
    ),
    SassVariableDeclaration(
      type: "SassVariableDeclaration",
      namespace: None,
      name: SassVariable(
        type: "SassVariable",
        name: Ident(
          type: "Ident",
          name: "my-single-item-list",
          raw: "my-single-item-list",
          span: Span(
            start: 34,
            end: 53,
          ),
        ),
        span: Span(
          start: 33,
          end: 53,
        ),
      ),
      value: SassList(
        type: "SassList",
        items: [
          Ident(
            type: "Ident",
            name: "alpha",
            raw: "alpha",
            span: Span(
              start: 56,
              end: 61,
            ),
          ),
        ],
        hasParens: true,
        separator: Comma,
        span: Span(
          start: 55,
          end: 63,
        ),
      ),
      important: None,
      overridable: false,
      forceGlobal: false,
      span: Span(
        start: 33,
        end: 64,
      ),
    ),
    SassVariableDeclaration(
      type: "SassVariableDeclaration",
      namespace: None,
      name: SassVariable(
        type: "SassVariable",
        name: Ident(
          type: "Ident",
          name: "box-shadow-first-two",
          raw: "box-shadow-first-two",
          span: Span(
            start: 66,
            end: 86,
          ),
        ),
        span: Span(
          start: 65,
          end: 86,
        ),
      ),
      value: SassList(
        type: "SassList",
        items: [
          SassList(
            type: "SassList",
            items: [
              Number(
                type: "Number",
                value: 0.0,
                raw: "0",
                span: Span(
                  start: 88,
                  end: 89,
                ),
              ),
              Length(
                type: "Length",
                value: Number(
                  type: "Number",
                  value: 0.5,
                  raw: "0.5",
                  span: Span(
                    start: 90,
                    end: 93,
                  ),
                ),
                unit: Ident(
                  type: "Ident",
                  name: "em",
                  raw: "em",
                  span: Span(
                    start: 93,
                    end: 95,
                  ),
                ),
                span: Span(
                  start: 90,
                  end: 95,
                ),
              ),
              SassVariable(
                type: "SassVariable",
                name: Ident(
                  type: "Ident",
                  name: "background",
                  raw: "background",
                  span: Span(
                    start: 97,
                    end: 107,
                  ),
                ),
                span: Span(
                  start: 96,
                  end: 107,
                ),
              ),
            ],
            hasParens: false,
            separator: Space,
            span: Span(
              start: 88,
              end: 107,
            ),
          ),
          SassList(
            type: "SassList",
            items: [
              Number(
                type: "Number",
                value: 0.0,
                raw: "0",
                span: Span(
                  start: 109,
                  end: 110,
                ),
              ),
              Length(
                type: "Length",
                value: Number(
                  type: "Number",
                  value: -0.5,
                  raw: "-0.5",
                  span: Span(
                    start: 111,
                    end: 115,
                  ),
                ),
                unit: Ident(
                  type: "Ident",
                  name: "em",
                  raw: "em",
                  span: Span(
                    start: 115,
                    end: 117,
                  ),
                ),
                span: Span(
                  start: 111,
                  end: 117,
                ),
              ),
              SassVariable(
                type: "SassVariable",
                name: Ident(
                  type: "Ident",
                  name: "background",
                  raw: "background",
                  span: Span(
                    start: 119,
                    end: 129,
                  ),
                ),
                span: Span(
                  start: 118,
                  end: 129,
                ),
              ),
            ],
            hasParens: false,
            separator: Space,
            span: Span(
              start: 109,
              end: 129,
            ),
          ),
        ],
        hasParens: false,
        separator: Comma,
        span: Span(
          start: 88,
          end: 129,
        ),
      ),
      important: None,
      overridable: false,
      forceGlobal: false,
      span: Span(
        start: 65,
        end: 130,
      ),
    ),
    SassVariableDeclaration(
      type: "SassVariableDeclaration",
      namespace: None,
      name: SassVariable(
        type: "SassVariable",
        name: Ident(
          type: "Ident",
          name: "list2",
          raw: "list2",
          span: Span(
            start: 132,
            end: 137,
          ),
        ),
        span: Span(
          start: 131,
          end: 137,
        ),
      ),
      value: SassList(
        type: "SassList",
        items: [
          SassList(
            type: "SassList",
            items: [
              Length(
                type: "Length",
                value: Number(
                  type: "Number",
                  value: 1.0,
                  raw: "1",
                  span: Span(
                    start: 140,
                    end: 141,
                  ),
                ),
                unit: Ident(
                  type: "Ident",
                  name: "px",
                  raw: "px",
                  span: Span(
                    start: 141,
                    end: 143,
                  ),
                ),
                span: Span(
                  start: 140,
                  end: 143,
                ),
              ),
              Ident(
                type: "Ident",
                name: "solid",
                raw: "solid",
                span: Span(
                  start: 144,
                  end: 149,
                ),
              ),
            ],
            hasParens: false,
            separator: Space,
            span: Span(
              start: 140,
              end: 149,
            ),
          ),
        ],
        hasParens: true,
        separator: Comma,
        span: Span(
          start: 139,
          end: 151,
        ),
      ),
      important: None,
      overridable: false,
      forceGlobal: false,
      span: Span(
        start: 131,
        end: 152,
      ),
    ),
  ],
  span: Span(
    start: 0,
    end: 153,
  ),
)
